cloudinit/distros/__init__.py

```diff
@@ -170,6 +170,7 @@ def __init__(self, name, cfg, paths):
         self.net_ops = iproute2.Iproute2
         self._runner = helpers.Runners(paths)
         self.package_managers: List[PackageManager] = []
+        self._selected_dhcp_client = None
```

Comment:
We may need to be of unpickling across upgrade paths as this class attribute won't exist in the pickled datasource and we are trying to reference it in the `dhcp_client` property below with self._selected_dhcp_client. I think we may need an _unpickle clause for this?

Reply:
> unpickling across upgrade paths

I'm fine with adding an attribute for now, but....

Long term why not just delete pickles on upgrade paths to avoid requiring workarounds like this - we do it across for Python upgrade path already so why not cloud-init? We have some pretty old attributes that we still check for that are very unlikely to ever be used again (see "fallback_nic").

Reply:
We can't blow away pickls across upgrade path because next boot of cloud-init will find no cached obj.pkl file which triggers cloud-init to rediscover all datasources again and recrawl the entire discovered IMDS which 1, would lead to longer reboot time and 2 expose the VM to not-detecting the datasource that was originally configured in cases where the IMDS is no longer available(sandboxed VMs) or non-responsive resulting in potentially destructive behavior like regenerating SSH host keys.

Reply:
>  1, would lead to longer reboot time 

@blackboxsw This would introduce a longer reboot once per quarter for distros that follow the upstream release cadence, that is only 4 times per year. All of the attribute checks at runtime aren't free either - once the pickle gets blown away the attribute checks wouldn't be required, so this would actually mean faster boots in the future by dropping unnecessary attribute checks when instantiating objects from pickles (which happens 3 times per boot on a fresh boot, 4 times per boot on a new boot). So this would lead to a longer reboot time _4 times a year_, and **shorter** reboot times _N-4 times per year (where N=number of reboots per year)_. The proposal isn't a net loss, it is a tradeoff that would also reduce the risk of introducing bugs due to missing attributes (like this would have been) and also reduce maintenance burden.


> 2 expose the VM to not-detecting the datasource that was originally configured

What new breakage would removing pickles at cloud-init version change introduce which doesn't already exist by removing pickles at Python version change? Surely if this is a supported use case of cloud-init, we must not depend _only_ on the existence of the pickle file but also have some other mechanism to prevent changing discovered datasources whenever a Python major version change happens under cloud-init? If we don't have some other mechanism and we say this is a supported use-case of cloud-init, then I would consider the lack of another mechanism to be a bug in cloud-init (i.e. we could, and probably should, change this regardless of whether we remove old caches on cloud-init upgrade).

Additionally, pickling was only added in release 20.4, ~3 years ago. Before the pickle behavior was added, wouldn't this have been the expected behavior across _every_ reboot (indeed, even every stage of cloud-init)?

Reply:
> The proposal isn't a net loss, it is a tradeoff that would also reduce the risk of introducing bugs due to missing attributes (like this would have been) and also reduce maintenance burden.

This argument is a bit of a stretch to me. A few attribute checks is going to be incredibly trivial to the overall run time, to the point where I don't think it's worth considering until we've taken much more drastic performance actions.

> What new breakage would removing pickles at cloud-init version change introduce which doesn't already exist by removing pickles at Python version change?

These are two very different upgrade use cases. A new (supported) Python version change is only going to be brought in by a do-release-upgrade or similar for other OSes. That's going to have to be intentionally triggered with ops monitoring and expected changes in behavior. A new version of cloud-init can be installed via unattended upgrade where unexpected changes are much more significant...though I agree that if we have bugs or unwieldly behavior, we should fix it.

Pickling issues have caused some bugs in the past, but our integration tests should handle them now.

That said, I do get that our pickling stuff is a big headache. It also seems less relevant as more clouds have moved towards wanting network data fetched every boot. I'd be open to alternative solutions, but I don't think that removing the pickle on every upgrade is it.